### PR TITLE
ng rebuild

### PR DESCRIPTION
**Pull request by sennui**
Known issues:
- [ ] - micro and key blocks re-use the same record 
- [ ] - still no beneficiary (next sprint)
- [ ] - no proof of fraud (next sprint)

Further issues to discuss:
- [ ] - default values for hash fields in blocks. Zeros may not be the best

### CODE REVIEW

File: apps/aecore/src/aec_conductor.erl [hunk 1]
```diff
@@ -632,16 +661,139 @@ handle_mining_reply({{error, {runtime, Reason}}, _}, State) ->
...
+                    seen_top_block_hash = SeenHash} = State) when MicroBlockHash =/= SeenHash ->
+    %% Candidate generated with stale top hash.
+    %% Regenerate the candidate.
+    State#state{micro_block_candidate = undefined};
+start_micro_signing(#state{consensus = #consensus{leader = true},
+                           micro_block_candidate = MicroCandidate} = State) ->
+    epoch_mining:info("Starting signing round"),
+    HeaderBin = MicroCandidate#candidate.bin,
+    Info      = [{top_block_hash, State#state.seen_top_block_hash}],
+    aec_events:publish(start_micro_signing, Info),
```
Comment (hanssv) [inline apps/aecore/src/aec_conductor.erl]: Who is listening to this?

File: apps/aecore/src/aec_txs_trees.erl [hunk 2]
```diff
@@ -30,8 +30,12 @@
 %%% API
 %%%===================================================================
 
--spec from_txs([aetx_sign:signed_tx()]) -> txs_tree().
-from_txs(Txs) ->
+-spec from_txs([aetx_sign:signed_tx(), ...] | []) -> txs_tree().
+
+from_txs([]) ->
+    %% NG-INFO: its fine to have empty transaction list (key-block)
```
Comment (hanssv) [inline apps/aecore/src/aec_txs_trees.erl]: Should go away

File: apps/aetx/src/aetx_sign.erl [hunk 3]
```diff
@@ -57,18 +60,24 @@ new(Tx, Signatures) ->
 %% @doc Given a transaction Tx, a private key or list of keys,
 %% return the cryptographically signed transaction using the default crypto
 %% parameters.
--spec sign(aetx:tx(), list(binary()) | binary()) -> signed_tx().
+-spec sign(aetx:tx()|binary(), list(binary()) | binary()) -> signed_tx() | tuple().
```
Comment (hanssv) [inline apps/aetx/src/aetx_sign.erl]: I don't like this spec-change... This function is a signing a transaction, not a general binary.

File: apps/aecore/src/aec_blocks.erl [hunk 4]
```diff
@@ -58,6 +65,12 @@ target(Block) ->
 difficulty(Block) ->
     aec_pow:target_to_difficulty(target(Block)).
 
+-spec is_key_block(block()) -> boolean().
+is_key_block(#block{miner = Miner, height = Height}) ->
```
[the remark this answers is not in the document]
Comment (sennui) [reply apps/aecore/src/aec_blocks.erl]: yes, it will into fall patterns

[the remark this answers is not in the document]
Comment (jur0) [reply apps/aecore/src/aec_blocks.erl]: Yes. There will be separate records for key and micro blocks.

File: apps/aetx/src/aetx_sign.erl (hunk 3 above)
Comment (jur0) [reply apps/aetx/src/aetx_sign.erl]: I will fix this. I can move signing of blocks into `aec_blocks` and the param will be a block, not a binary.

File: apps/aecore/src/aec_conductor.erl [hunk 5]
```diff
@@ -177,7 +184,7 @@ handle_call({add_synced_block, Block},_From, State) ->
     {reply, Reply, State1};
 handle_call(get_block_candidate,_From, State) ->
     Res =
-        case State#state.block_candidate of
+        case State#state.micro_block_candidate of
```
[the remark this answers is not in the document]
Comment (sennui) [reply apps/aecore/src/aec_conductor.erl]: true, only makes sense when we are a leader. New check would improve API

**Pull request closed without merging**